engine/src/main/java/org/terasology/engine/rendering/primitives/ChunkMeshImpl.java

```diff
@@ -0,0 +1,273 @@
+     */
+    @Override
+    public void discardData() {
+        vertexElements = null;
+    }
+
+    private void renderVbo(int id) {
+        if (vertexBuffers[id] <= 0 || disposed) {
+            return;
+        }
+
+        GL30.glBindVertexArray(vaoCount[id]);
+        GL30.glDrawElements(GL30.GL_TRIANGLES, vertexCount[id], GL30.GL_UNSIGNED_INT, 0);
+        GL30.glBindVertexArray(0);
+    }
+
+    /**
+     * Updates a given material with information such as the World position of a chunk and whether it is animated.
+     *
+     * @param chunkMaterial a Material instance to be updated
+     * @param chunkPosition a Vector3f instance holding the world coordinates of a chunk
+     * @param chunkIsAnimated a boolean: true if the chunk is animated, false otherwise
+     */
+    @Override
+    public void updateMaterial(Material chunkMaterial, Vector3fc chunkPosition, boolean chunkIsAnimated) {
+        chunkMaterial.setFloat3("chunkPositionWorld", chunkPosition, true);
+        chunkMaterial.setFloat("animated", chunkIsAnimated ? 1.0f : 0.0f, true);
+    }
+
+    public int render(ChunkMesh.RenderPhase type) {
+        switch (type) {
+            case OPAQUE:
+                renderVbo(0);
+                break;
+            case ALPHA_REJECT:
+                renderVbo(1);
+                renderVbo(2);
+                break;
+            case REFRACTIVE:
+                renderVbo(3);
+                break;
+            default:
+                break;
+        }
+        return triangleCount();
+    }
+
+    /**
+     * Disposes of all the data stored in an instance of this class and the associated data stored in the GLBufferPool instance provided on
+     * construction.
+     * <p>
+     * ChunkMesh instances cannot be un-disposed.
+     */
+    @Override
+    public void dispose() {
+        if (!disposed) {
+            disposalHook().dispose();
+            for (int i = 0; i < vertexBuffers.length; i++) {
+                int id = vertexBuffers[i];
+                if (id != 0) {
+                    GL30.glDeleteBuffers(id);
+                    vertexBuffers[i] = 0;
+                }
+
+                id = idxBuffers[i];
+                if (id != 0) {
+                    GL30.glDeleteBuffers(id);
+                    idxBuffers[i] = 0;
+                }
+
+                id = vaoCount[i];
+                if (id != 0) {
+                    GL30.glDeleteVertexArrays(id);
+                    vaoCount[i] = 0;
+                }
+            }
+
+            disposed = true;
+        }
+    }
+
+    @Override
+    public DisposableHook disposalHook() {
+        WeakReference<ChunkMeshImpl> chunkMeshWeakReference = new WeakReference<>(this);
+        return new DisposableHook() {
+            private final int[] hookVertexBuffers = vertexBuffers;
+            private final int[] hookIndexBuffers = idxBuffers;
+            private final int[] hookVertexArrayObjects = vaoCount;
```

Comment:
Weak reference to `this`, but direct references to private properties of this? Suspicious.

Reply:
that's because its getting wrapped into a lambda

Reply:
What is because what is getting wrapped in what?

Reply:
this would be in the context of DiposableHook when the context is in the class that is creating the hook.

Reply:
As I read this, it looks like ChunkMesh has both a `dispose` method that's visible to the `engine.rendering.primitives` package, as well as this public `disposalHook` method.

And ChunkMeshImpl's implementation of `dispose` uses `disposalHook` (creating a new instance and then immediately invoking it), but **also** does all the same operations itself?

This implementation of `DisposableHook` is written so it _does_ hold references to ChunkMeshImpl's private array fields—this makes sense if there's something holding on to these ChunkMeshImpl.DisposableHooks beyond the time the ChunkMeshImpl objects themselves have been garbage collected, and indeed that's what the ChunkMesh interface's javadoc says, but I'm clearly missing something. What is building and saving this list of DisposableHooks and why is it important for it to do that instead of using ChunkMesh.dispose?

And can we get rid of the code duplication between ChunkMeshImpl.dispose and ChunkMeshImpl.disposalHook().dispose(), or are my sleepy eyes reading it wrong and it's not actually a duplication?

Reply:
It used for entities that are associated with a chunkMesh but I can't see it being used anywhere.

Reply:
addressed by f392965a7c931867e11eaa8a94f5a5dcb073afa2